### PR TITLE
Improve typing of dynamic method lookup for keyboard handlers

### PR DESCRIPTION
**Pull request by lukemelia**
Thanks @alex-cardstack for the assist!

### CODE REVIEW

**Comment by lukemelia** [comment]
@alex-cardstack our work type-checked but didn't actually work because `Object.keys` does not include keys on prototypes. I pushed a commit to address this but couldn't figure out how to do it without one unfortunate cast to `Record<string, unknown>`. Maybe you have an idea?

File: packages/boxel/addon/components/boxel/input/time/keyboard.ts [hunk 1]
```diff
@@ -59,6 +58,27 @@ abstract class BaseKeyboardHandler implements KeyboardHandler {
     this.component.setMeridian('pm');
   }
 
+  private handlerMethodFromEventCode(eventCode: string) {
+    type ValidStringKey = Extract<keyof this, string>;
+
+    const isValidMethodKey = (
+      methodName: string
+    ): methodName is ValidStringKey => {
+      return getMethodNames(this as Record<string, unknown>).includes(
```
Comment (danibonilha) [inline packages/boxel/addon/components/boxel/input/time/keyboard.ts:67]: nit: you could omit the return keyword

File: packages/boxel/addon/utils/reflection.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    .concat(deepFunctions(Object.getPrototypeOf(obj)) || []);
+};
+
+const distinctDeepFunctions = (obj: Record<string, unknown>) =>
+  Array.from(new Set(deepFunctions(obj)));
+
+export const getMethodNames = (obj: Record<string, unknown>): string[] =>
+  distinctDeepFunctions(obj).filter(
+    (name: string) => name !== 'constructor' && !~name.indexOf('__')
+  );
```
Comment (danibonilha) [inline packages/boxel/addon/utils/reflection.ts:20]: you could use a generic type here, then you don't need to cast it: 
```ts
export const getMethodNames = <Obj extends Record<keyof Obj, unknown>>(
  obj: Obj
): string[] =>
```

File: packages/boxel/addon/utils/reflection.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const isGetter = (obj: unknown, name: string): boolean =>
+  !!(Object.getOwnPropertyDescriptor(obj, name) || {}).get;
+
+const isFunction = (obj: Record<string, unknown>, name: string): boolean =>
```
Comment (danibonilha) [inline packages/boxel/addon/utils/reflection.ts:4]: Maybe we could create a type with  `Record<string, unknown>` to reuse it

File: packages/boxel/addon/utils/reflection.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+const isGetter = (obj: unknown, name: string): boolean =>
```
Comment (danibonilha) [inline packages/boxel/addon/utils/reflection.ts:1]: isn't this object also `Record<string, unknown>` ?

File: packages/boxel/addon/utils/reflection.ts (hunk 2 above)
Comment (lukemelia) [reply packages/boxel/addon/utils/reflection.ts:20]: Yay, this is the concept I was missing!

**Review by danibonilha — APPROVED** [review]
LGTM

**Pull request merged**